pkg/metrics/metric_sample_pool.go

```diff
@@ -0,0 +1,37 @@
+package metrics
+
+import (
+	"sync"
+)
+
+// MetricSamplePool is a pool of metrics sample
+type MetricSamplePool struct {
+	pool *sync.Pool
+}
+
+// NewMetricSamplePool creates a new MetricSamplePool
+func NewMetricSamplePool(batchSize int) *MetricSamplePool {
+	return &MetricSamplePool{
+		pool: &sync.Pool{
+			New: func() interface{} {
+				return make([]MetricSample, batchSize)
+			},
+		},
+	}
+}
+
+// GetBatch gets a batch of metric samples from the pool
+func (m *MetricSamplePool) GetBatch() []MetricSample {
+	if m == nil {
+		return nil
+	}
+	return m.pool.Get().([]MetricSample)
+}
+
+// PutBatch puts a batch back into the pool
+func (m *MetricSamplePool) PutBatch(batch []MetricSample) {
+	if m == nil {
+		return
+	}
+	m.pool.Put(batch[:cap(batch)])
```

Comment:
Currently the pool gives out slice batches with a length equal to the slice cap. That way the pool users can assign `len(batch)` elements. We could give slices of length 0 and cap N but it would force the users to use `cap` and check that they don't `append` out of bounds to avoid re-allocating.

Reply:
Got it, nice one.